euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,385 @@
+|Split eigenvectors|(n_gamma_points, 3*n_ions, n_ions, 3) array of complex floats|
+|Gamma point indices|(n_gamma_points,) array of ints| 
+|Energy units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Calculate crystal structure factor (see [algorithms - 2.1](04_algorithms.md))|Neutron scattering lengths, temperature, Debye-Waller exponent (optional)|Crystal structure factor (discrete energies)|
+|Calculate Debye-Waller exponent (see [algorithms - 3.2](04_algorithms.md))|Temperature|Debye-Waller exponent|
+|Calculate density of states|Energy bins, energy broadening|1D Spectrum object|
+|Serialise|Filename|File|
+
+### Crystal Structure Factor (discrete energies)
+
+This contains the structure factor calculated for a crystal at each q-point and
+for each mode. It could be passed to Brille or Horace for energy binning and
+plotting, or binned and plotted in Euphonic.
+
+|Attribute|Type|
+|---------|----|
+|Q-points|(n_qpts, 3) array of floats|
+|Structure factor|(n_qpts, 3*n_ions) array of floats|
+|Frequencies|(n_qpts, 3*n_ions) array of floats|
+|Energy units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Calculate binned crystal structure factor (see [algorithms - 2.2](04_algorithms.md))|Energy bins, temperature (optional)|2D Spectrum object|
+|Serialise|Filename|File|
+
+### Debye-Waller exponent
+
+The Debye-Waller exponent is calculated as a sum over the 1st BZ, so requires
+frequencies and eigenvectors that have been calculated on a grid. It is also
+direction dependent, and resolved per ion, so will be a 3x3 array for each ion
+in the structure.
+
+It is used as an input for the structure factor calculations, this could be
+optional but is needed if the user wants to include temperature effects.
+
+|Attribute|Type|
+|---------|----|
+|Structure|Structure object|
+|Temperature|float|
+|Debye-waller|(n_ions, 3, 3) array of floats|
+|Temperature units|string|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Serialise|Filename|File
+
+### 2D Spectrum
+
+This is an object for any 2D binned spectra, and can be plotted or exported to
+another program. It is intended to be general and can be used for binned crystal
+Q-E plots, powder |Q|-E plots etc. The x-tick locations/labels have been
+included to allow helpful labelling of the x-axis, which is most often the
+Q-axis (e.g. high symmetry points), and as the structure/q-point information is
+not included in the 2D spectrum object these cannot be calculated on the fly.
+For parity it could be sensible to include y tick labels/locations, but until
+Q-Q plots are a Euphonic requirement, these are not included in the current
+object definition.
+
+|Attribute|Type|
+|---------|----|
+|x bin edges|(n_xbins + 1) array of floats|
+|x bin units|string|
+|y bin edges|(n_ybins + 1) array of floats|
+|y bin units|string|
+|x tick labels|list of strings|
+|x tick locations|array of ints|
```

Comment:
If the x tick locations/labels are always used together to set special points, maybe they should be a single attribute: a list of 2-tuples `[(loc1, label1), (loc2, label2), ...]`. We should _not_ use a dict of `{label: loc}` because repeated labels are quite common. `{loc: label}` would also be unsafe as multiple labels might be set at the same x-position to represent jumps from one special point to another. And then if the attribute is set to `None` we know to fall back on default x-labelling based on q values; it's not obvious what should happen if *x_tick_locations=[1, 2 3]* and *x_tick_labels=None*.

An edge case we hit in Sumo is that you can't even rely on a label always referring to the same vector in **q**-space; between symmetry and periodicity it might make sense for a path to cover more than one equivalent special point.

Reply:
Are x-tick locations int values to point number or a position? 
Python `zip` would neatly join the two lists of labels+locations for use in a loop or something if two lists are easier to manage than a list of tuples.

Reply:
`zip` won't complain if the lists are different lengths, it will quietly skip the excess entries.

Reply:
x-tick locations `int` values are to point to a bin number.

I agree that a list of tuples would be better than 2 lists, I'll update this

Reply:
Is that the index of a bin _edge_ or a bin _midpoint_?

Reply:
Bin midpoint, as for the use case of labelling high-symmetry points there would be 1 q-point per bin and you would want the label to be at the middle of that bin.

However, if the axis were an energy axis, it's probably more likely you would want to label bin edges, but I imagine this use case would be covered by Matplotlib's default tick labels? I can't really think of a better way to handle this...